feat(log-viewer): added log viewer

mattnolting opened the pull request:
closes #3811

Preview: https://patternfly-pr-4029.surge.sh/components/log-viewer
Designs here: https://marvelapp.com/prototype/bia3j2d
React issue: https://github.com/patternfly/patternfly-react/issues/5341

patternfly-build commented:
Preview: https://patternfly-pr-4029.surge.sh

mattnolting commented:
@arivepr Updated

nicolethoen reviewed (approved):
it definitely looks beautiful! 
I'm happy to work with you and Alexander to incorporate these.

mcarrano reviewed (commented):
This is looking good @mattnolting . Just a couple of questions...

* I wasn't exactly sure about this example

![Screen Shot 2021-04-29 at 11 28 32 AM](https://user-images.githubusercontent.com/6619860/116577113-32000400-a8de-11eb-966d-d8628972e6dd.png)

Does this represent what the user would see as the result of a find? If so, can we highlight matching strings vs whole rows? Also, wasn't sure where the lime green color comes from.

* For the responsive layout, the way elements collapse looks good. Should we have an example with the search field opened?

mcarrano commented:
@mattnolting after consulting with @mceledonia we decided to use the following colors to represent "found" items. All rows that match are highlighted using `Gold-50` and the current selected row is highlighted with `Gold-100`. Here's what that should look like:

![Logviewer-find](https://user-images.githubusercontent.com/6619860/116718098-9fc43280-a9a7-11eb-9d49-af4d452ee063.png)

nicolethoen commented:
@mcarrano What is the use case for selecting a logger row? I can understand hover styles being used to help make reading the log viewer a little easier, and the highlighting of found search results, but what is a 'selected row' representing?

mcarrano commented:
@nicolethoen In the find feature, you may have hits on multiple rows. But selected on mean the one you are currently focuses on. The up/down arrows in the Search input let you move that focus. "Selected" might not be the best term, but not sure what to call that.

kmcfaul commented:
Maybe `matched row` would make sense for that modifier then?

Everything looks good to me, just looks like the highlighted colors need to be updated to the golds.

jessiehuff reviewed (commented):
This looks pretty awesome, @mattnolting!!

I think the structure looks pretty good, and I like the approach. The only aspect that I'm not positive I understand the exact interaction yet is the search feature and found/selected results. Is it going to behave more like the Typeahead Select where using the arrow keys just changes visual focus and not keyboard focus? (When you're in typeahead, browser focus stays in the input so that you can continue typing/interacting with the input.) Is that similar to how this would be handled?

Also, currently every log item has a tabIndex which makes me wonder if there will ever be interactive content in the log viewer. Will it always be static items or will there ever be the expectation to interact with log viewer items?

DaoDaoNoCode reviewed (commented):
@mattnolting Looks great, nice work! A small question, why we need both `With drilldown expanded` and `With dropdown expanded` examples? I know the first one is for mobile view and the second one is for desktop view, but they will not show at the same time, right? Will it be better to use a single example named like `With dropdown/drilldown menu open` instead of these 2 to show dropdown menu on desktop and show drilldown menu on mobile? It will more clear to me, because currently the `With drilldown expanded` example is the same as the basic example on large screen width, and `With dropdown expanded` example is the same as the basic example on small screen width. I cannot differentiate them from basic example at the first glance, which may also confuse users. Just an advice! Other parts look perfect, thanks!

DaoDaoNoCode commented:
BTW, I found a small issue on `Search Input`. In the design, the left icon is "search previous" and the right icon is "search next".

<img width="348" alt="Screen Shot 2021-05-03 at 3 36 16 PM" src="https://user-images.githubusercontent.com/37624318/116925808-ac56be00-ac27-11eb-81ea-6d06391a7d7f.png">

However, in the implement of `Search Input`, they are different.

<img width="280" alt="Screen Shot 2021-05-03 at 3 31 27 PM" src="https://user-images.githubusercontent.com/37624318/116925912-d01a0400-ac27-11eb-9591-8314a79fda74.png">

Which makes this example look weird IMO.
@mattnolting @mcarrano WDYT? The problem is not in this PR, should we create an issue to track and fix this in `Search Input` component to make it the same as the design? Thanks.

mcarrano commented:
Good catch @DaoDaoNoCode . Yes that looks like a bug in the HTML example. In the React example they are correct:

![Screen Shot 2021-05-03 at 5 00 46 PM](https://user-images.githubusercontent.com/6619860/116933247-4bcc7e80-ac31-11eb-9cde-6ddef4beb983.png)

mattnolting commented:
Great feedback @DaoDaoNoCode! I've combined the examples with expanded dropdowns per your suggestion. 

@mcarrano @kmcfaul I've updated `pf-m-selected` to `pf-m-match` and `pf-m-current`. 

@mcarrano I've updated the hover style to match `pf-m-match` and @jessiehuff  removed `tabindex` on `pf-c-log-viewer__list-item`.

The `pf-c-search-input` navigation buttons have been updated as well, which will affect both `pf-c-search-input` and `pf-c-log-viewer`.

DaoDaoNoCode reviewed (approved):
LPTM!! Thanks @mattnolting 👏

patternfly-build commented:
:tada: This PR is included in version 4.103.0 :tada:

The release is available on:
- [GitHub release](https://github.com/patternfly/patternfly/releases/tag/prerelease-v4.103.0)
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@patternfly/patternfly/v/4.103.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

dlabrecq reviewed (commented):
LGTM
